packages/gatsby/src/redux/run-sift.js

```diff
@@ -242,6 +259,70 @@ const getBucketsForQueryFilter = (
   return true
 }
 
+/**
+ * @param {string} typedKey
+ * @param {IDbQueryElemMatch} filter
+ * @param {Array<string>} nodeTypeNames
+ * @param {FiltersCache} filtersCache
+ * @param {Array<FilterCache>} filterCaches Matching node sets are put in this array
+ */
+const collectBucketForElemMatch = (
+  typedKey,
+  filter,
+  nodeTypeNames,
+  filtersCache,
+  filterCaches
+) => {
+  let comparator = ``
+  let targetValue = null
+  let f /*: DbQuery*/ = filter
+  while (f) {
+    if (f.type === `elemMatch`) {
+      const q /*: IDbQueryElemMatch */ = f
+      f = q.nestedQuery
+    } else {
+      const q /*: IDbQueryQuery */ = f
+      comparator = q.query.comparator
+      targetValue = q.query.value
+      break
+    }
+  }
+
+  if (
+    ![
+      `$eq`,
+      // "$lte",
+      // "$gte",
+    ].includes(comparator)
+  ) {
+    return false
+  }
+
+  if (!filtersCache.has(typedKey)) {
+    ensureIndexByElemMatch(typedKey, filter, nodeTypeNames, filterCaches)
+  }
+
+  const nodesByKeyValue /*: Set<IGatsbyNode> | undefined*/ = getFilterCacheByTypedChain(
+    typedKey,
+    comparator,
+    targetValue,
+    filtersCache
+  )
+
+  // If we couldn't find the needle then maybe sift can, for example if the
+  // schema contained a proxy; `slug: String @proxy(from: "slugInternal")`
+  // There are also cases (and tests) where id exists with a different type
+  if (!nodesByKeyValue) {
```

Comment:
How certain are we that it's not an empty Set? My paranoia kinda wants to add a `filterCaches.size` check here too but maybe we want empty Sets to be added.

Reply:
A `Set` is only created when actually adding a value to the cache, not prematurely. So if a set exists in the (FilterCache) cache, then there must have been at least one value for it

Reply:
I should rename this var btw, it should be `filterCache`.

Hmmm, that made me realize you might actually be correct. This isn't a `Set<IGatsbyNode>` (which has this guarantee) but rather a `FilterCache`, which is created prematurely and which may end up with an empty `Set` of nodes.... nice catch!

Reply:
Scratch that, this _is_ the `Set<IGatsbyNode> | undefined`, so it does have this property. I just don't know a better name for it without getting into Java territory.